File: tools/cp-caps/test_csl.py
#!/usr/bin/env python3
#
#  Copyright (c) 2025, The OpenThread Authors.
#  All rights reserved.
#
#  Redistribution and use in source and binary forms, with or without
#  modification, are permitted provided that the following conditions are met:
#  1. Redistributions of source code must retain the above copyright
#     notice, this list of conditions and the following disclaimer.
#  2. Redistributions in binary form must reproduce the above copyright
#     notice, this list of conditions and the following disclaimer in the
#     documentation and/or other materials provided with the distribution.
#  3. Neither the name of the copyright holder nor the
#     names of its contributors may be used to endorse or promote products
#     derived from this software without specific prior written permission.
#
#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
#  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
#  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
#  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
#  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
#  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
#  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
#  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
#  POSSIBILITY OF SUCH DAMAGE.
#

import unittest

from otci import OTCI
from device_manager import DeviceManager


class TestCsl(unittest.TestCase):
    """Test whether the DUT supports the CSL feature."""

    @classmethod
    def setUpClass(cls):
        cls.__device_manager = DeviceManager()
        cls.__dut = cls.__device_manager.dut
        cls.__ref = cls.__device_manager.ref

    def setUp(self):
        self.__dut.factory_reset()
        self.__ref.factory_reset()

    def test_csl_transmitter(self):
        """Test whether the DUT supports the CSL transmitter."""
        packets = 10
        dataset = self.__device_manager.get_default_dataset()

        self.__dut.join(dataset)
        self.__dut.wait_for('state', 'leader')

        # Set the reference device as an SSED
        self.__ref.set_mode('-')
        self.__ref.config_csl(channel=15, period=320000, timeout=100)
        self.__ref.join(dataset)
        self.__ref.wait_for('state', 'child')

        child_table = self.__dut.get_child_table()
        ret = len(child_table) == 1 and child_table[1]['csl']

        if ret:
            ref_mleid = self.__ref.get_ipaddr_mleid()
            result = self.__dut.ping(ref_mleid, count=packets, interval=1)
            ret = result['transmitted_packets'] == result['received_packets'] == packets

        self.__dut.leave()
        self.__ref.leave()

        self.__device_manager.output_test_result_bool('CSL Transmitter', ret)

    def tearDown(self):
        pass

    @classmethod
    def tearDownClass(cls):
        cls.__dut.close()
        cls.__ref.close()


if __name__ == '__main__':
    unittest.main()

File: tests/nexus/test_csl.cpp
/*
 *  Copyright (c) 2026, The OpenThread Authors.
 *  All rights reserved.
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions are met:
 *  1. Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *  2. Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in the
 *     documentation and/or other materials provided with the distribution.
 *  3. Neither the name of the copyright holder nor the
 *     names of its contributors may be used to endorse or promote products
 *     derived from this software without specific prior written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 *  POSSIBILITY OF SUCH DAMAGE.
 */

#include <stdio.h>

#include "platform/nexus_core.hpp"
#include "platform/nexus_node.hpp"

namespace ot {
namespace Nexus {

static constexpr uint32_t kFormNetworkTime     = 13 * 1000;
static constexpr uint32_t kAttachAsSsedTime    = 20 * 1000;
static constexpr uint32_t kCslPeriodUs         = 288000;
static constexpr uint32_t kCslPeriod           = kCslPeriodUs / OT_US_PER_TEN_SYMBOLS;
static constexpr uint32_t kCslSyncTime         = 10 * 1000;
static constexpr uint16_t kEchoPayloadSize     = 10;
static constexpr uint32_t kEchoResponseTimeout = 3 * 1000;
static constexpr uint32_t kWaitTimeAfterPing   = 1 * 1000;

void TestCsl(void)
{
    Core nexus;

    Node &parent = nexus.CreateNode();
    Node &ssed   = nexus.CreateNode();

    parent.SetName("PARENT");
    ssed.SetName("SSED");

    nexus.AdvanceTime(0);

    SuccessOrQuit(Instance::SetGlobalLogLevel(kLogLevelDebg));

    AllowLinkBetween(parent, ssed);

    Log("---------------------------------------------------------------------------------------");
    Log("Step 1: Form network with Parent Leader and attach SSED child");

    parent.Form();
    nexus.AdvanceTime(kFormNetworkTime);
    VerifyOrQuit(parent.Get<Mle::Mle>().IsLeader());

    ssed.Join(parent, Node::kAsSed);
    nexus.AdvanceTime(kAttachAsSsedTime);
    VerifyOrQuit(ssed.Get<Mle::Mle>().IsAttached());

    ssed.Get<Mac::Mac>().SetCslPeriod(kCslPeriod);
    nexus.AdvanceTime(kCslSyncTime);
    VerifyOrQuit(ssed.Get<Mac::Mac>().IsCslEnabled());

    Log("---------------------------------------------------------------------------------------");
    Log("Step 2: SSED pings Parent");

    nexus.SendAndVerifyEchoRequest(ssed, parent.Get<Mle::Mle>().GetMeshLocalEid(), kEchoPayloadSize,
                                  Ip6::kDefaultHopLimit, kEchoResponseTimeout);
    nexus.AdvanceTime(kWaitTimeAfterPing);

    Log("---------------------------------------------------------------------------------------");
    Log("Step 3: Parent pings SSED");

    nexus.SendAndVerifyEchoRequest(parent, ssed.Get<Mle::Mle>().GetMeshLocalEid(), kEchoPayloadSize,
                                  Ip6::kDefaultHopLimit, kEchoResponseTimeout);
    nexus.AdvanceTime(kWaitTimeAfterPing);
}

} // namespace Nexus
} // namespace ot

int main(void)
{
    ot::Nexus::TestCsl();
    printf("All tests passed\n");
    return 0;
}
